docs/_posts/2021-03-29-using-seahorn.md

```diff
@@ -0,0 +1,331 @@
+<!-- more sophisticated optimizations and to disable SSE and AVX features. -->
+
+<!-- ``` -->
+<!-- export RUSTFLAGS="-Copt-level=1 $RUSTFLAGS" -->
+<!-- export RUSTFLAGS="-Cno-vectorize-loops -Cno-vectorize-slp $RUSTFLAGS" -->
+<!-- export RUSTFLAGS="-Ctarget-feature=-mmx,-sse,-sse2,-sse3,-ssse3,-sse4.1,-sse4.2,-3dnow,-3dnowa,-avx,-avx2 $RUSTFLAGS" -->
+<!-- ``` -->
+
+<!-- (This is not needed for our simple example.) -->
+
+
+## Running SeaHorn
+
+Having built a bitcode file containing the program and all the libraries that it
+depends on, we now need to fix a few things in the bitcode file to make it
+suitable for [SeaHorn].
+This includes deleting the `main` function rustc added, replacing panic handling
+(calling `__VERIFIER_error` instead), and removing the body of `_eprint` and
+`_print` (the internal implementations of the printing macros).
+
+The `rvt-patch-llvm` tool can do all this for us:
+
+``` shell
+rvt-patch-llvm -o try_seahorn.patch.bc --seahorn -vv target/debug/deps/try_seahorn-*.bc
+```
+
+The bitcode file `try_seahorn.patch.bc` is now ready to be processed by
+[SeaHorn].
+
+To run [SeaHorn] we use a configuration file that we checkout from
+[verify-c-common].
+The Docker image is configured so that `SEAHORN_VERIFY_C_COMMON_DIR` points to
+that checkout.
+In addition, because we deleted the `main` function rustc added to the bitcode
+(which does some runtime initialisation and then calls the original main
+function) we have to tell [SeaHorn] the name of the entry function.
+To find this name execute this:
+
+``` shell
+llvm-nm --defined-only try_seahorn.patch.bc | grep main | cut -d ' ' -f3
+```
+
+The output should look something like this:
+
+``` shell
+_ZN11try_seahorn4main17h8733453d83f64f5aE
+```
+
+(the suffix after "17h" might be different)
+
+We can now put everything together and run [SeaHorn] like this:
+
+``` shell
+sea yama -y ${SEAHORN_VERIFY_C_COMMON_DIR}/seahorn/sea_base.yaml bpf --entry="_ZN11try_seahorn4main17h8733453d83f64f5aE" try_seahorn.patch.bc
+```
+
+This command will produce a long output that should end like this:
+
+```
+sat
+
+
+************** BRUNCH STATS ***************** 
+[...]
+************** BRUNCH STATS END ***************** 
+```
+
+The "sat" indicates that [SeaHorn] was able to find an execution of the program
+that violates the assertion (or panics).
+[SeaHorn] can produce a trace of the execution like this:
+
+``` shell
+sea yama -y ${SEAHORN_VERIFY_C_COMMON_DIR}/seahorn/sea_cex_base.yaml bpf --entry="_ZN11try_seahorn4main17h8733453d83f64f5aE" try_seahorn.patch.bc
+```
+
+As [SeaHorn] is an LLVM-based tool, the trace is an LLVM IR trace.
+Moreover, [SeaHorn] does a few transformations and optimisations to the program,
+and the trace is over the resulting program.
+Fortunately, [SeaHorn] preserves some debug information that allows it to print
+source code line-numbers along the trace.
+
+Scroll up the trace until you find this:
```

Comment:
Is it not possible to execute the trace and either:

- get a stackdump when it crashes - pointing me at the source line that failed

or

- run it in a debugger and use breakpoints on error reporting functions to make it stop at the failure point so that I can see the source line number (from the Rust code) that it failed on, inspect local variables, look at a stack dump (again with Rust source files/lines)

??

If I can't do either of these, then, when I am using SeaHorn with some other program, how would I know what to look for in the LLVM IR trace file that it produces?